dspy/utils/callback.py

```diff
@@ -256,60 +256,96 @@ def on_evaluate_end(
 
 
 def with_callbacks(fn):
-    @functools.wraps(fn)
-    def wrapper(instance, *args, **kwargs):
-        # Combine global and local (per-instance) callbacks.
-        callbacks = dspy.settings.get("callbacks", []) + getattr(instance, "callbacks", [])
-
-        # If no callbacks are provided, just call the function
-        if not callbacks:
-            return fn(instance, *args, **kwargs)
-
-        # Generate call ID as the unique identifier for the call, this is useful for instrumentation.
-        call_id = uuid.uuid4().hex
+    """Decorator to add callback functionality to instance methods."""
 
+    def _execute_start_callbacks(instance, fn, call_id, callbacks, args, kwargs):
```

Comment:
nit: since they are internal methods within with_callbacks, we don't need to add "_" prefix.